core/src/main/scala/org/apache/spark/storage/BlockId.scala

```diff
@@ -87,6 +87,29 @@ case class ShufflePushBlockId(shuffleId: Int, mapIndex: Int, reduceId: Int) exte
   override def name: String = "shufflePush_" + shuffleId + "_" + mapIndex + "_" + reduceId
 }
 
+@DeveloperApi
+case class ShuffleMergedBlockId(appId: String, shuffleId: Int, reduceId: Int) extends BlockId {
+  override def name: String = "mergedShuffle_" + appId + "_" + shuffleId + "_" + reduceId + ".data"
+}
+
+@DeveloperApi
+case class ShuffleMergedIndexBlockId(
+  appId: String,
+  shuffleId: Int,
+  reduceId: Int) extends BlockId {
+  override def name: String =
+    "mergedShuffle_" + appId + "_" + shuffleId + "_" + reduceId + ".index"
```

Comment:
`appId` would not be sufficient - we can have multiple attempts for an application.
(here and in other block id's introduced).

Reply:
The merged shuffle files are already written by the server and the file identifier doesn't include the attemptId. To include attemptId we do require a change on the push side as well where attempt Id is also included in the PushBlockMessage and the server uses that to create merged files. I think this is a much bigger change and we have this jira [SPARK-32923 ](https://issues.apache.org/jira/browse/SPARK-32923) to address it.

One thing we can do here is that we introduce attemptId in these new Ids that we are creating but not use them while trying to find file and add a TODO.
What do you think @mridulm @Victsm @zhouyejoe ?

Reply:
Sorry, I thought this comment was about adding stage attempt Id. However, this is about application attempt Id which is not covered by SPARK-32923.

Reply:
This should be sufficient as there would be only one attempt running at the same time. If one attempt fails to run, the local directories will be removed. So even though different attempts are using the same tmp path, it won't be problematic.

Reply:
The scenario I am thinking of is as follows:

app1-atttempt1 fails - but due to some factor (overloaded RM, out-of-band messages, etc), ESS on nodes which are not running containers for app1-attempt1 are not immediately notified.
app1-attempt2 starts and subsequently starts shuffling data.
Assume very quick tasks/bootstrap/etc for simplification just for the scenario.

In this case, can there be a problem ? ESS writes to a merge directory which is going to get deleted when app1-attempt1 failure is finally communicated to it ?


+CC @tgravescs who has much more context about these things than I do :-)

Reply:
This is a valid point. I think we can add UUID.randomUUID to the merge folder path, same as the original shuffle local files dirs.

Reply:
This is not going to be quite simple . If we attach UUID.randomUUID to merge dir path, then there are 2 major issues with that:
1. We want a single merge directory under an application location dir. When a new executor starts, it checks whether the merge_manager exists. It only creates it if the merge_manager doesn't exist. If there is a random UUID attached, then how to let the new executors know when **not** to create the new directory? For blockMgr dirs this is not a requirement and each executor creates its own directory.

2. How does the server know the exact name of merge_manager directory which has a random UUID? The server currently constructs out the merge_manager path from the `local_dirs` in `ExecutorShuffleInfo`. It actually finds the parent dir of the block manager `local_dirs` and then assumes that `merge_manager` exists under it. The code is here:
https://github.com/apache/spark/blob/12abfe79173f6ab00b3341f3b31cad5aa26aa6e4/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java#L1003
This would require sending the merge_manager directory name to the server as well. This requires updating `RegisterExecutor` message but since we can't change an existing message, we will have to create a new one.

I still think we can ignore this edge case because if the merged files are lost there will be a fallback to original blocks. There are 2 cases:
1. Server hasn't finalized a shuffle yet and merged folder is deleted. When the shuffle is finalized, since the files are not there, the partition will not be considered merged and so original shuffle blocks will be fetched.
2. Server has finalized and merged folder is deleted after that. In this case, during fetch of the merged partition by client, the server will respond with failure. This will cause the client to fallback on the original un-merged blocks.

@zhouyejoe @mridulm Let me know what you think

Reply:
@otterc If the merge_dir can get created under the blockmgr_UUID.randomUUID dir? Since RegisterExecutor message would send this blockmgr_UUID.randomUUID dir to ESS, ESS would know what are the local dirs to be used for merge dir. In our internal version, ESS will use the first RegisterExecutor message as the merge dirs list to be used. Suppose the scenario described: Executor1 gets the local dirs list "/grid/[a-c]/yarn/usercache/test/appcache/application_id/", Executor2 gets the local dirs list "/grid/[d-f]". Either of them will create the merge_dir under their own local dirs, for example: Executor1 creates "/grid/[a-c]/........./merge_dir"  and Executor2 creates "/grid/[d-f]//........./merge_dir". Executor1 is lucky to be the the first one to register with the local ESS, ESS will only use the dirs for "/grid/[a-c]". But during the Executor registration, the actual dirs in the message is "/grid/a/yarn/tmp/usercache/testuser/appCache/application_id/blockmgr_RandomID", "/grid/b/yarn/tmp/usercache/testuser/appCache/application_id/blockmgr_RandomID",
"/grid/c/yarn/tmp/usercache/testuser/appCache/application_id/blockmgr_RandomID". Internally, we trim the dirs to the "/grid/[a_c]/yarn/tmp/usercache/testuser/appCache/application_id/", assume the merge_dirs are "/grid/[a_c]/yarn/tmp/usercache/testuser/appCache/application_id/mergedirs". Even though there are other "/grid/[d-f]//........./merge_dir" created by Executor2, but since ESS will only use the dirs from the Executor1, those dirs will not be used. With the same logic, if we move merge_dir to subdir of blockmgr_RandomIDs, this will still work, right? The only cons would be there would be empty merge_dirs/subdirs created by other executors.

Reply:
@zhouyejoe As I mentioned earlier, this edge case where merge_manager directory gets deleted, is handled by the fallback. The client will fetch original blocks instead of merged blocks. merge directory is not related to a block manager directory so why should it be created under it? If the goal is just to handle this edge case then this doesn't seem to be a clean solution.
What is the benefit of making this change which is not clean when we can fallback for these exceptional cases?

Reply:
Thanks for the review and discussion on this issue. We have reached an agreement that the current implementation will cause issues when the application has multiple attempts. 
**Issue description:**
Currently the merge dir is created under the application local tmp dir by the executors. For example, for application_12345, executor1 running on Node1 will create the merge_dir under /[yarn-local-dirs]/usercache/[username]/appcache/application_12345/merge_dir. Even though there are other executors from this application running on Node1, they will not create those dirs as it has been created. When executor1 register itself with the shuffle service on Node1, it will also register the local dirs and let shuffle service know where the unmerged shuffle data is stored. So in this case, it will register  /[yarn-local-dirs]/usercache/[username]/appcache/application_12345/blockmgr_RandomID. ESS will use this dirs and assume the merge dirs have been created under /[yarn-local-dirs]/usercache/[username]/appcache/application_12345/merge_dir. For the executor register messages from the same application, the first message will register the merge dirs in shuffle service, others will be ignored. While the application finishes, we rely on Yarn NodeManager to clean up the whole dir /[yarn-local-dirs]/usercache/[username]/appcache/application_12345. 
1nd issue: Since shuffle service will only use the first executor register message from the same application, the original dirs list from the first executor in the first attempt will be kept all the way during the later attempts run. 
2st issue: While in the case of multiple attempts from the same application, this dir won’t get cleaned up. So the original merged shuffle files generated by shuffle service from the former attempts will still exist within the merge_dir. This would be stale data but shuffle service will use these stale data to serve the merge shuffle fetch in later attempts.

**Root cause of the issue:**
For original unmerged shuffle files, there won’t be such issues as 1. Every executor will register its own local dirs to shuffle service 2. Block manager dirs are encoded with randomIDs on the path. For merged shuffle dir, we only use the first ever register executor message to get the merge_dir base location and ignore all the later ones. But in executor register message, there is no attemptID encoded, so shuffle service cannot get to know whether they are from different attempts. Yarn will not cleanup tmp directories while the application is still running, even though there are multiple attempts.

**Proposed solution:**
In either proposed solution, we have to get the attemptID encoded into the executor register protocol. After we have the attemptID encoded and shuffle service can use this attemptID as the way to distinguish among multiple attempts.
Solution 1: We can create a new protocol for RegisterExecutor, which adds a new fields: AttemptID, compared with the original RegisterExecutor protocol. This new protocol will be used when push based shuffle is enabled.
Solution 2: Don’t create new protocol and just leverage the existing protocol. We can encode “appID” + “attemptID” as the “new appID” in the RegisterExecutor, let’s call it appAttemptID. Shuffle service will be able to tell whether the String is the original appID or appAttemptID with Regex, and then deal with it.

CC @Ngone51 @tgravescs @attilapiros. Can you share your opinions over the issues and proposed solutions? Thanks.

Reply:
To minimize change, why not get the executor to create the directory as "merge_directory_<attempt_id>" ? (if missing)
While registering, ESS can pick up the highest attempt id and associate the executor with that attempt id (and so merge_directory).

Thoughts ?

Reply:
Yes, I think this would be a simpler option without changing the existing protocol.
But I found it is not so straight forward to get the attemptID within the Executor.

Reply:
This is a good point, unfortunately we do not set the attempt id in spark conf like we do for app id.
A few options here would be:

a) Also propagate attempt id via "spark.app.attemptId" is available (and use default value if missing).
b) Defer registeration/directory creation to first task being run : ShuffleMapTask has `appAttemptId` as part of it. 
c) For yarn, CONTAINER_ID env variable can be parsed to fetch attempt id - though this might not be optimal.

I am not very keen on modifying protocol if possible.

Thoughts ?

Reply:
+CC @tgravescs for suggestions.

Reply:
Below are my comments for each solution 
1. Create a new `RegisterExecutor` message. If we do this, we might also want to add the recent "merge_manager_attemptx` name. A benefit of that would be to remove the logic on the server which tries to find the merge_manager directory path by finding the parent of blockMgrs directory.

2. Encoding `attemptId` in the `appId` of `RegisterExecutor` message. The old `ExternalShuffleBlockResolver` needs to know the `appId` (without attemptId) when the executors are registered and the `executors` map is populated, since all the `get...BlockData()` apis reference this map with only `appId` and `executorId`. We can't change these apis. `RemoteBlockPushResolver` may also need to know the `appId` and `attemptId` separately to avoid changes to multiple data-structures. It would be then better to handle this parsing in `ExternalBlockHandler`. 

3. Have the `RemoteBlockPushResolver` figure out the attemptId just by itself by listing the directory and finding the latest attempt. With this one every time an executor registers, the remote shuffle has to do the following:
- find the parent directory from blockManager dirs. 
- list the merge_manager_* dirs  
- figure out the latest attempt.
All these 3 steps above are avoided in solution 2 for every executor registration because we know the attempt Id from the message itself and it will ignore the ones belonging to the current one. The additional cost that is incurred in solution 2 is just splitting the `appId` and `attemptId` which I don't think is that much.

I think Solution 1 is the cleanest but I can see that it adds another message and attemptId may be just specific to Yarn. Between 2 and 3, I prefer solution 2.

WDYT @tgravescs @mridulm @attilapiros @Ngone51 @Victsm @zhouyejoe

Reply:
I have a very tricky idea here, which is based on solution 3: 

When an executor of an application attempt X tries to create the merge directory, it could first check whether the dir `merge_manager_X-1` exists. If exists, delete it. And if dir `merge_manager_X` not exists, then create `merge_manager_X`. And if the executor becomes the one who creates the merge dir, we send the `ExecutorShuffleInfo` with the special `shuffleManager`, e.g., "sort_merge_manager_attemptX". And `ExternalBlockHandler` can parse the `shuffleManager` into two parts. So `ExternalShuffleBlockResolver` can still register the normal `ExecutorShuffleInfo` and  `RemoteBlockPushResolver` can know whether to update the merge dir.


Besides, I'd prefer soultion1. I think it's reasonable to add a new message for the push-based shuffle. We can have a new type for it, e.g.,`RegisterMergeDirectory`, which includes the merge directory directly and attempted too (of course).

Reply:
@Ngone51 Moving the management to executor is a nice idea - If we are moving the delete to the executor, then we could formulate it to make the change minimal right ?

In executor: 
a) List application/merge_manager_*
b) Delete all merge_manager_* directories [1] which are older than the current attempt id [2]
b.1) If there are newer merge_manager_* directories - it exit's : since this is not the latest attempt for this app (should be vanishingly small chance - but adding for completeness sake).
c) Create merge_manager_$ATTEMPT_ID, and register with existing rpc's.

In shuffle service:
d) Shuffle service simply looks up the latest merge_manager_* (which should ideally be a single directory), and registers application/executor with that attempt.
d.1) If earlier attempts exist, cleanup metadata - directory would be cleaned up already anyway.

Thoughts ?

[1] Given possibility of concurrent executors deleting, we should be robust to the errors there.
[2] In theory, there could be race with a newer version - but should not happen.
 - that is, if merge_manager_attemptId

Reply:
cc @Victsm

Reply:
Btw, forgot to add my thoughts on a new rpc message.
I am referring to adding something like `RegisterExecutorForPushBasedShuffle` (or some such) in addition to existing `RegisterExecutor`. That is, send `RegisterExecutorForPushBasedShuffle` after `RegisterExecutor` succeeds.
 
A few thoughts:

* If ESS does not support the new RPC, how is the spark application supposed to behave ?
  * Case where `RegisterExecutor` would succeed while `RegisterExecutorForPushBasedShuffle` would fail due to `IllegalArgumentException` at ESS due to unrecognized msg id.
  * Currently, it would throw a `SparkException` and leads to executor failure : do we change this behavior ? Or do we simply fail the application due to unsupported config ?
* Adding the new rpc allows us to decouple executor registration from whether the executor host should be candidate for hosting mergers or not.
  * This will help with future evolution.
* If we are taking this path, it would be better for ESS to manage the merger location entirely - and not have executors create/update it (as discussed above). It will help ESS evolve independently.

Reply:
> If ESS does not support the new RPC, how is the spark application supposed to behave ?

So this is the case that the cluster does not support push-based shuffle but client is sending push-based shuffle related messages. In this case, it should be okay that the executor fails early because it is trying push-based shuffle which will not work. For this case, we can throw a `SparkException` with message that "push-based is not supported by the cluster so please turn it off".

> If we are taking this path, it would be better for ESS to manage the merger location entirely - and not have executors create/update it (as discussed above). It will help ESS evolve independently.

ESS cannot create the merge_manager directory under application local directory because it doesn't have permissions to do so. App local dirs have permission `750` and shuffle service is part of NM process that are run usually by `yarn` user.

Reply:
> If we are moving the delete to the executor, then we could formulate it to make the change minimal right ?

Yes, and thanks for bringing the detailed steps, which looks good.


BTW, there's another tricky way to ship the merge directory that is appending the merge directory into `ExecutorShuffleInfo.localDirs`.

Also, adding a new RPC doesn't seem to help a lot here according to the discussion.

Reply:
> ESS cannot create the merge_manager directory under application local directory because it doesn't have permissions to do so.

Good point, thanks.

Reply:
> Also, adding a new RPC doesn't seem to help a lot here according to the discussion.

I agree, minimizing change to protocol should be better here.
Let us try without introducing RPC - it does not allow for clean separation of concerns anyway (based on @otterc's clarification above) ... so making it entirely managed at executors would be better.

Reply:
Thanks for all the discussion above. I will update the PR with the proposed ideas above: 1. No new RPC needs to be introduced. 2. The dirs will still be created by Executor as of the the permission issue. 3. Will let the Executor manage the merge_dir creation. 4. Will let the Executor to delete the last attempt's created merge_dir if it exists. 5. Register the merge attempt Dirs through ExecutorShuffleInfo, where a new String for shuffleManager field to distinguish the attemptID, e.g., "sort_merge_manager_attemptX", suggested by @Ngone51 . 

@mridulm  IIUC, if we have the item 5 there, we don't need the shuffle service to list the dirs and figure out the largest attemptIDs for merge dirs, right?

Reply:
You can always have a race between creation and deletion, right ?
a) Executor e1.attempt2 checks - found no directories.
b) Executor e2.attempt1 checks - found no directories
c) Both create directories.
d) Both register to ESS - now ESS sees two directories.

Reply:
@zhouyejoe If we are sending the merge directory name via RegisterExecutor message which is the 5th step, then I don't think 4th step is necessary where the executor deletes directory of old attempt.
Currently, the executors don't delete the previous previous attempts block manager dirs and neither does the shuffle service. It is deleted by Yarn when the application is finally finished. All the blockMgr dirs hang around until the app is done.
With merge_manager directory it will be the same case. In this solution, the remote shuffle service doesn't depend on  listing the dirs so what do we get from having the executor delete old attempt merge_manager dir?

Reply:
To add to my comment above (should have provided more context): I was assuming we are relying on directory existence to infer latest.
But I like @Ngone51's idea here better:
>  And if the executor becomes the one who creates the merge dir, we send the ExecutorShuffleInfo with the special shuffleManager, e.g., "sort_merge_manager_attemptX". And ExternalBlockHandler can parse the shuffleManager into two parts

`shuffleManager` should to be relevant when we had multiple shuffle managers - and from evolution point of view, we could always introduce a new shuffle in future.
My proposal is extension of the idea above - here, attempt is metadata about shuffleManager we want to convey.
Ideally, this should be within `ExecutorShuffleInfo` - but given the compatibility issues here, why not add a general way to encode metadata about the shuffle manager here ?

That is, instead of hardcoding 'SortShuffleManager_attemptX' and 'SortShuffleManager' as supported patterns, we could simply allow for 'SortShuffleManager' (existing) and 'SortShuffleManager:json_string' ?
This can allow for any potential future evolution to also be possible - with currently merge_dir: "merge_directory_<attemptid>" being the initial field (@zhouyejoe let us explicitly pass the directory name, instead of inferring it from attempt id ?)

Ofcourse if ':' is missing in shuffleManager, then we treat it as empty metadata case.

With this in place, we would still need changes to push block protocol to include attempt id - but given no one is using that currently, we can make that change.
With this explicit specification of attempt id by executor, we remove all ambiguity.

Thoughts @Ngone51 ?

Reply:
Updated accordingly. Please help reivew. @mridulm @Ngone51 @otterc .

Reply:
This will be handled as part of SPARK-30602, resolving thread.